File: backend/mcp-code-intelligence-python/src/mcp_code_intel/memory/embedding/factory.py
"""Factory for creating EmbeddingService — tries Ollama, then ONNX local."""

import sys
from pathlib import Path
from typing import Any

from .provider import EmbeddingProvider
from .onnx_provider import OnnxEmbeddingProvider
from .ollama_provider import OllamaEmbeddingProvider
from .service import EmbeddingService


class EmbeddingFactory:
    """Create EmbeddingService: Ollama (if configured) → ONNX local → None."""

    @staticmethod
    def create(config: dict[str, Any], vector_repo) -> EmbeddingService | None:
        """Try providers in priority order, return first available."""
        service = EmbeddingFactory._try_ollama(config, vector_repo)
        if service:
            return service
        service = EmbeddingFactory._try_onnx(config, vector_repo)
        if service:
            return service
        return None

    @staticmethod
    def _try_ollama(config: dict, vector_repo) -> EmbeddingService | None:
        """Attempt Ollama provider if URL is configured."""
        from ...ollama import create_client

        client = create_client(config)
        if not client:
            return None
        model = config.get("ollama_model", "nomic-embed-text")
        provider = OllamaEmbeddingProvider(client, model)
        _log(f"Embedding: Ollama ({model})")
        return EmbeddingService(provider, vector_repo)

    @staticmethod
    def _try_onnx(config: dict, vector_repo) -> EmbeddingService | None:
        """Attempt local ONNX provider if model files exist."""
        workspace = config.get("workspace", "")
        model_path = EmbeddingFactory._resolve_model(workspace)
        vocab_path = EmbeddingFactory._resolve_vocab(workspace)
        if not model_path or not vocab_path:
            _log("ONNX model not found. Place model.onnx + vocab.txt in .code-intel/models/")
            return None
        try:
            provider = OnnxEmbeddingProvider(model_path, vocab_path)
            if not provider.is_available():
                return None
            _log("Embedding: ONNX local (all-MiniLM-L6-v2, 384d)")
            return EmbeddingService(provider, vector_repo)
        except Exception as e:
            _log(f"ONNX init failed: {e}")
            return None

    @staticmethod
    def _resolve_model(workspace: str) -> Path | None:
        """Find model.onnx in workspace or home .code-intel/models/."""
        candidates = [
            Path(workspace) / ".code-intel" / "models" / "model.onnx",
            Path(workspace) / ".code-intel" / "models" / "all-MiniLM-L6-v2.onnx",
            Path.home() / ".code-intel" / "models" / "model.onnx",
        ]
        return next((p for p in candidates if p.exists()), None)

    @staticmethod
    def _resolve_vocab(workspace: str) -> Path | None:
        """Find vocab.txt in workspace or home .code-intel/models/."""
        candidates = [
            Path(workspace) / ".code-intel" / "models" / "vocab.txt",
            Path.home() / ".code-intel" / "models" / "vocab.txt",
        ]
        return next((p for p in candidates if p.exists()), None)


def _log(msg: str) -> None:
    print(f"[embed-factory] {msg}", file=sys.stderr, flush=True)

File: backend/mcp-code-intelligence-nodejs/src/memory/embedding/factory.ts
/**
 * EmbeddingFactory — creates EmbeddingService with priority: Ollama → ONNX → null.
 * Matches Python/Kotlin behavior for provider selection.
 */

import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import { EmbeddingProvider } from './provider.js';
import { OnnxEmbeddingProvider } from './onnx-provider.js';
import { OllamaEmbeddingProvider, OllamaClient } from './ollama-provider.js';
import { EmbeddingService } from './service.js';
import { VectorRepository } from '../vector-repo.js';

/** Config shape expected by factory. */
interface EmbeddingConfig {
  workspace: string;
  ollamaUrl?: string | null;
  ollamaModel?: string;
}

export class EmbeddingFactory {
  /** Try providers in priority order, return first available. */
  static create(config: EmbeddingConfig, vectorRepo: VectorRepository): EmbeddingService | null {
    const service = EmbeddingFactory.tryOllama(config, vectorRepo);
    if (service) return service;
    const onnxService = EmbeddingFactory.tryOnnx(config, vectorRepo);
    if (onnxService) return onnxService;
    return null;
  }

  /** Attempt Ollama provider if URL is configured. */
  private static tryOllama(config: EmbeddingConfig, vectorRepo: VectorRepository): EmbeddingService | null {
    if (!config.ollamaUrl) return null;
    const model = config.ollamaModel ?? 'nomic-embed-text';
    const client = createOllamaClient(config.ollamaUrl, model);
    if (!client.isAvailable()) return null;
    const provider = new OllamaEmbeddingProvider(client, model);
    log(`Embedding: Ollama (${model})`);
    return new EmbeddingService(provider, vectorRepo);
  }

  /** Attempt local ONNX provider if model files exist. */
  private static tryOnnx(config: EmbeddingConfig, vectorRepo: VectorRepository): EmbeddingService | null {
    const modelPath = EmbeddingFactory.resolveModel(config.workspace);
    const vocabPath = EmbeddingFactory.resolveVocab(config.workspace);
    if (!modelPath || !vocabPath) {
      log('ONNX model not found. Place model.onnx + vocab.txt in .code-intel/models/');
      return null;
    }
    try {
      const provider = new OnnxEmbeddingProvider(modelPath, vocabPath);
      if (!provider.isAvailable()) return null;
      log('Embedding: ONNX local (all-MiniLM-L6-v2, 384d)');
      return new EmbeddingService(provider, vectorRepo);
    } catch (e: any) {
      log(`ONNX init failed: ${e.message}`);
      return null;
    }
  }

  /** Find model.onnx in workspace or home .code-intel/models/. */
  private static resolveModel(workspace: string): string | null {
    const candidates = [
      path.join(workspace, '.code-intel', 'models', 'model.onnx'),
      path.join(workspace, '.code-intel', 'models', 'all-MiniLM-L6-v2.onnx'),
      path.join(os.homedir(), '.code-intel', 'models', 'model.onnx'),
    ];
    return candidates.find(p => fs.existsSync(p)) ?? null;
  }

  /** Find vocab.txt in workspace or home .code-intel/models/. */
  private static resolveVocab(workspace: string): string | null {
    const candidates = [
      path.join(workspace, '.code-intel', 'models', 'vocab.txt'),
      path.join(os.homedir(), '.code-intel', 'models', 'vocab.txt'),
    ];
    return candidates.find(p => fs.existsSync(p)) ?? null;
  }
}

/** Create a minimal Ollama HTTP client for embeddings. */
function createOllamaClient(baseUrl: string, model: string): OllamaClient {
  return {
    async getEmbedding(text: string): Promise<number[] | null> {
      try {
        const url = `${baseUrl}/api/embeddings`;
        const resp = await fetch(url, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ model, prompt: text }),
        });
        if (!resp.ok) return null;
        const data = await resp.json() as { embedding?: number[] };
        return data.embedding ?? null;
      } catch {
        return null;
      }
    },
    isAvailable(): boolean {
      // Synchronous check — assume available if URL is configured
      return !!baseUrl;
    },
  };
}

function log(msg: string): void {
  process.stderr.write(`[embed-factory] ${msg}\n`);
}
